loki/subroutine.py

```diff
@@ -131,31 +131,21 @@ def __setstate__(self, s):
         # Ensure that we are attaching all symbols to the newly create ``self``.
         self.rescope_symbols()
 
-    @staticmethod
-    def _infer_allocatable_shapes(spec, body):
+
+    def _infer_allocatable_shapes(self):
         """
         Infer variable symbol shapes from allocations of ``allocatable`` arrays.
         """
-        alloc_map = {}
-        for alloc in FindNodes(ir.Allocation).visit(body):
+        for alloc in FindNodes(ir.Allocation).visit(self.body):
             for v in alloc.variables:
                 if isinstance(v, sym.Array):
                     if alloc.data_source:
-                        alloc_map[v.name.lower()] = alloc.data_source.type.shape
+                        new_shape = alloc.data_source.type.shape
                     else:
-                        alloc_map[v.name.lower()] = v.dimensions
-        vmap = {}
-        for v in FindVariables().visit(body):
-            if v.name.lower() in alloc_map:
-                vtype = v.type.clone(shape=alloc_map[v.name.lower()])
-                vmap[v] = v.clone(type=vtype)
-        smap = {}
-        for v in FindVariables().visit(spec):
-            if v.name.lower() in alloc_map:
-                vtype = v.type.clone(shape=alloc_map[v.name.lower()])
-                smap[v] = v.clone(type=vtype)
-        return (SubstituteExpressions(smap, invalidate_source=False).visit(spec),
-                SubstituteExpressions(vmap, invalidate_source=False).visit(body))
+                        new_shape = v.dimensions
+
+                    # Clone the symbol with the updated type to inject shape info
+                    v.clone(type=v.type.clone(shape=new_shape))
```

Comment:
Any reason for not using `v.type = v.type.clone(shape=new_shape)` here? Seems like the canonical API use

Reply:
Good catch! Updated now (sorry, habitual rebase).